### DISCUSSION TITLE
Multiple mg_http_serve_dir at different URLs

### DISCUSSION THREAD

**Discussion by Nandaleio** [cesanta/mongoose#2973] [Ask questions, get answers] [answered]
I was wondering if it would be possible to serve multiples directories at different URLs 

I have this ev_handler function in my `main.c` :

```C
#define MG_API_HEADERS "Content-Type: application/json\r\nAccess-Control-Allow-Origin: *\r\nAccess-Control-Allow-Methods: *\r\n"

static void ev_handler(struct mg_connection *c, int ev, void *ev_data) {
    if (ev != MG_EV_HTTP_MSG) return;

    struct mg_http_message *hm = (struct mg_http_message *) ev_data;

    if (mg_strcmp(hm->method, mg_str("OPTIONS")) == 0) {
            return mg_http_reply(c, 204, MG_API_HEADERS"Access-Control-Allow-Headers: *\r\n\r\n", "{\"res\": \"No Content\"}");
    } 

    if(mg_match(hm->uri, mg_str("/api/#"), NULL)) {
        return mg_http_reply(c, 200, MG_API_HEADERS, "{\"test\": \"%s\"}\n", "test");
    }

    if(mg_match(hm->uri, mg_str("/admin/#"), NULL)) {
        struct mg_http_serve_opts opts = {
            .root_dir = "/ui/dist",
            .fs = &mg_fs_packed
        };
        return mg_http_serve_dir(c, ev_data, &opts);
    }
    else {
        struct mg_http_serve_opts opts = {
            .root_dir = "./public",
            
        };
        return mg_http_serve_dir(c, hm, &opts);
    }
}
```

Whenever I try to reach `http://localhost:8080/admin/` I get a `404 Not Found`  error
I tried to swap the `opts` of the `if(mg_match(hm->uri, mg_str("/admin/#"), NULL)) {` ans the `else {` but whatever is in the first one gets the 404 error.

Do you know if it's possible to achieve this ? 
Am I using the right method ?

Note that I'm using Embedded filesystem in one of the `mg_http_serve_dir`
Also I'm using `MG_VERSION "7.15"`  

Thank you !

**Comment by scaprile** [comment, marked as answer]
Please see our [documentation](https://mongoose.ws/documentation/),  and follow the guidelines in our [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples).

Nothing prevents that
https://mongoose.ws/documentation/tutorials/http/http-server/
https://mongoose.ws/documentation/#mg_http_serve_dir
https://mongoose.ws/documentation/#struct-mg_http_serve_opts

https://gurubase.io/g/mongoose/mongoose-serve-multiple-directories-different-urls

other path:
https://mongoose.ws/documentation/#mg_match

https://gurubase.io/g/mongoose/serve-multiple-directories-with-mg-match

**Reply by Nandaleio** [reply]
I tried to change my code to the following like in the links you send me :


```C
// HTTP event handler
static void ev_handler(struct mg_connection *c, int ev, void *ev_data) {
    if (ev != MG_EV_HTTP_MSG) return;

    struct mg_http_message *hm = (struct mg_http_message *) ev_data;

    struct mg_str caps[1];

    struct mg_http_serve_opts opts;
    memset(&opts, 0, sizeof(opts));

    // Check which pattern the URI matches
    if (mg_match(hm->uri, mg_str("/admin"), caps)) {
        printf("admin-");
        opts.root_dir = "./public";
    } else if (mg_match(hm->uri, mg_str("/nyan"), caps)) {
        printf("nyan-");
        opts.root_dir = "./test";
    } else if (mg_match(hm->uri, mg_str("/docs"), caps)) {
        printf("docs-");
        opts.root_dir = "/ui/dist";
        opts.fs = &mg_fs_packed;
    } else {
      // Default to a 404 response if no pattern matches
      mg_http_reply(c, 404, "Content-Type: text/plain\r\n", "TEST TEST TEST");
      return;
    }
    
    printf("serv: %s\r\n", opts.root_dir);
    mg_http_serve_dir(c, hm, &opts);
}
```

Even though the `mg_match` works fine now, I still cannot access any files. The only endpoint accessible is the else part `mg_http_reply(c, 404, "Content-Type: text/plain\r\n", "TEST TEST TEST");`

The example shown here https://gurubase.io/g/mongoose/serve-multiple-directories-with-mg-match doesn't seems to work (at least with my code) 

I don't understand why it behave that way. The documentation doesn't show that case.

Do you know where this problem can come from ?

**Reply by scaprile** [reply]
Sorry, I won't check your code, you have lots of examples, plenty of docs, and a tutorial doing exactly what you want, think you did not follow:
https://mongoose.ws/documentation/tutorials/http/http-server/#dynamic-restful-server
```
    if (mg_match(hm->uri, mg_str("/api/sum"), NULL)) {
```
If your printf is not executed, it is because there is no match.
If there is no match, it is because either
a) the argument does not match
b) you are not calling mg_match() properly

**Reply by Nandaleio** [reply]
The printf is executed but I still get error 404

**Reply by scaprile** [reply]
No, it is not, and if it is, then your flow is wrong
the 404 is on the `else` how can that possibly execute after any of the "then" ? That is basic C

And BTW, the AI example is using wildcards, you are not. Not saying that you need them, just saying that you broke it.

**Comment by Nandaleio** [comment]
I think I found a way to achieve what I wanted thanks to @smilediver and https://github.com/cesanta/mongoose/issues/1289

I'm not sure how but changing the uri of the mg_http_message fix the error 404

Here the `+= 6` is because of the number of char after the first `/` of the url in the `mg_match` and beside the wildcard `#`

```C
if(mg_match(hm->uri, mg_str("/admin/#"), NULL)) {
        hm->uri.buf += 6;
        hm->uri.len -= 6;
        opts.root_dir = "./public";
    } else {
        opts.root_dir = "/ui/dist";
        opts.fs = &mg_fs_packed;
    }
    
    mg_http_serve_dir(c, hm, &opts);
```

and match the base url (here `/admin/`) with base href of the html page

```html
  <base href="/admin/">
```

**Comment by scaprile** [comment]
What you want can be perfectly done if you just read the docs and do as the examples suggest
